File: ios/HybridMixer.swift
import CoreMedia
import Foundation
import NitroModules
import QuartzCore

/// The `Mixer` HybridObject: the scene,
/// the output settings, the compositor (`VideoMixer`) and both encoders.
///
/// Threading: JS calls that touch the scene are posted to the render queue;
/// `video`/`audio`/`stats` use a lock. The H.264 encoder is fed from the
/// render queue, the AAC encoder from the microphone queue (under
/// `audioLock`), and both encoders hand their output to the session's sink.
final class HybridMixer: HybridMixerSpec {
  let videoMixer = VideoMixer()

  private let lock = NSLock()
  private var videoSettings = VideoSettings(width: 720, height: 1280, frameRate: 30, bitrateKbps: 2500,
                                            keyframeIntervalSeconds: 2)
  private var audioSettings = AudioSettings(sampleRate: 48000, channels: 1, bitrateKbps: 128)
  private var errorCallback: ((CaptureError) -> Void)?
  private var encoding = false
  private var encodedVideoFrames: UInt64 = 0
  private var encodedAudioFrames: UInt64 = 0
  private var encoderFailures: UInt64 = 0
  private var outputWindow: [(time: CFTimeInterval, bytes: Int)] = []  // last second of encoder output
  private var cameraLayerAttached = false
  private var videoStarted = false
  private var epoch: CFTimeInterval?  // the capture time of the first frame handed to the encoders
  private weak var sink: EncodedFrameSink?
  private var pendingAudio: [(adts: Data, time: CFTimeInterval)] = []  // until the first video frame is out
  private static let maxPendingAudio = 100  // about 2 s of AAC packets

  // Render queue only.
  private var h264: H264Encoder?

  // Microphone queue and start/stop, under audioLock.
  private let audioLock = NSLock()
  private var aac: AacEncoder?
  private var microphone: HybridMicrophoneSource?

  override init() {
    super.init()
    let settings = videoSettings
    videoMixer.queue.async { [weak self] in
      self?.videoMixer.setOutput(width: Int(settings.width), height: Int(settings.height), frameRate: settings.frameRate)
    }
  }

  func dispose() {
    stopEncoding()
    videoMixer.queue.async { [weak self] in
      guard let self else { return }
      for layer in self.videoMixer.layers {
        self.detach(layer)
      }
      self.videoMixer.layers.forEach { self.videoMixer.removeLayer($0.object) }
    }
    setAudioSourceInternal(nil)
  }

  // MARK: - HybridMixerSpec

  var video: VideoSettings {
    get {
      lock.lock()
      defer { lock.unlock() }
      return videoSettings
    }
    set {
      lock.lock()
      videoSettings = newValue  // applied at the next startEncoding when already encoding
      let apply = !encoding
      lock.unlock()
      if apply {
        videoMixer.queue.async { [weak self] in
          self?.videoMixer.setOutput(width: Int(newValue.width), height: Int(newValue.height),
                                     frameRate: newValue.frameRate)
        }
      }
    }
  }

  var audio: AudioSettings {
    get {
      lock.lock()
      defer { lock.unlock() }
      return audioSettings
    }
    set {
      lock.lock()
      audioSettings = newValue
      lock.unlock()
    }
  }

  var stats: MixerStats {
    let (captured, rendered, dropped) = videoMixer.counters
    lock.lock()
    let now = CACurrentMediaTime()
    let bytes = outputWindow.filter { now - $0.time <= 1 }.reduce(0) { $0 + $1.bytes }
    let stats = MixerStats(
      capturedFrames: Double(captured), renderedFrames: Double(rendered), droppedFrames: Double(dropped),
      encodedVideoFrames: Double(encodedVideoFrames), encodedAudioFrames: Double(encodedAudioFrames),
      videoBitrateKbps: Double(bytes) * 8 / 1000, encoderFailures: Double(encoderFailures))
    lock.unlock()
    return stats
  }

  var isEncoding: Bool {
    lock.lock()
    defer { lock.unlock() }
    return encoding
  }

  func addLayer(layer: any HybridVideoLayerSpec, frame: LayerFrame?) throws {
    guard let content = content(of: layer) else { return }
    let scene = SceneLayer(content: content, frame: frame)
    videoMixer.queue.async { [weak self] in
      guard let self else { return }
      guard !self.videoMixer.layers.contains(where: { $0.object === scene.object }) else { return }
      self.attach(scene)
      self.videoMixer.addLayer(scene)
      self.updateCameraLayerFlag()
    }
  }

  func removeLayer(layer: any HybridVideoLayerSpec) throws {
    let object = layer as AnyObject
    videoMixer.queue.async { [weak self] in
      guard let self else { return }
      if let scene = self.videoMixer.layers.first(where: { $0.object === object }) {
        self.detach(scene)
      }
      self.videoMixer.removeLayer(object)
      self.updateCameraLayerFlag()
    }
  }

  func setLayerFrame(layer: any HybridVideoLayerSpec, frame: LayerFrame) throws {
    let object = layer as AnyObject
    videoMixer.queue.async { [weak self] in self?.videoMixer.setFrame(frame, of: object) }
  }

  func setAudioSource(source: (any HybridMicrophoneSourceSpec)?) throws {
    setAudioSourceInternal(source as? HybridMicrophoneSource)
  }

  func onError(callback: @escaping (CaptureError) -> Void) throws {
    lock.lock()
    errorCallback = callback
    lock.unlock()
  }

  // MARK: - Session link (called by HybridRtmpPublisher on its queue)

  /// Creates both encoders with the current settings and starts feeding them.
  /// Idempotent. The first video frame is forced to be a keyframe.
  func startEncoding(sink: EncodedFrameSink) {
    videoMixer.queue.async { [weak self] in
      guard let self else { return }
      self.lock.lock()
      let alreadyEncoding = self.encoding
      let videoSettings = self.videoSettings
      let audioSettings = self.audioSettings
      self.lock.unlock()
      guard !alreadyEncoding else { return }

      guard self.videoMixer.isAvailable, let encoder = H264Encoder(settings: videoSettings) else {
        self.reportEncoderFailure("could not create the H.264 encoder for \(Int(videoSettings.width))x\(Int(videoSettings.height))")
        return
      }
      encoder.sink = sink
      encoder.onOutput = { [weak self] bytes in self?.videoFrameEncoded(bytes: bytes) }
      encoder.onFailure = { [weak self] message in self?.reportEncoderFailure(message) }
      self.h264 = encoder

      let aac = AacEncoder(settings: audioSettings)
      aac?.onPacket = { [weak self] adts, time in self?.audioPacketEncoded(adts, time: time) }
      aac?.onFailure = { [weak self] message in self?.reportEncoderFailure(message) }
      if aac == nil {
        self.reportEncoderFailure("could not create the AAC encoder for \(Int(audioSettings.sampleRate)) Hz")
      }
      self.audioLock.lock()
      self.aac = aac
      self.audioLock.unlock()

      self.lock.lock()
      self.encoding = true
      self.sink = sink
      self.epoch = nil
      self.videoStarted = false
      self.pendingAudio.removeAll()
      self.outputWindow.removeAll()
      self.lock.unlock()

      self.videoMixer.setOutput(width: Int(videoSettings.width), height: Int(videoSettings.height),
                                frameRate: videoSettings.frameRate)
      self.videoMixer.setEncodeHandler { [weak self] pixelBuffer, time in
        guard let self, let h264 = self.h264 else { return false }
        return h264.encode(pixelBuffer: pixelBuffer, ptsMs: self.ptsMs(for: time))
      }
    }
  }

  /// Stops feeding the encoders and invalidates them. Idempotent.
  func stopEncoding() {
    videoMixer.queue.async { [weak self] in
      guard let self else { return }
      self.videoMixer.setEncodeHandler(nil)
      self.h264?.invalidate()
      self.h264 = nil
      self.audioLock.lock()
      self.aac = nil
      self.audioLock.unlock()
      self.lock.lock()
      self.encoding = false
      self.sink = nil
      self.pendingAudio.removeAll()
      self.lock.unlock()
    }
  }

  /// Microphone queue: encode synchronously; `audioPacketEncoded` holds
  /// the packets back until the first video frame is out.
  func audioArrived(_ sampleBuffer: CMSampleBuffer, muted: Bool) {
    audioLock.lock()
    defer { audioLock.unlock() }
    aac?.encode(sampleBuffer, muted: muted)
  }

  /// Capture reconfiguration failures are separate from encoder failures.
  func reportCaptureFailure(_ failure: CaptureFailure) {
    lock.lock()
    let callback = errorCallback
    lock.unlock()
    callback?(CaptureError(code: failure.code, message: failure.message))
  }

  // MARK: - Internals

  private func content(of layer: any HybridVideoLayerSpec) -> SceneLayer.Content? {
    if let camera = layer as? HybridCameraSource { return .camera(camera) }
    if let image = layer as? HybridImageLayer { return .image(image) }
    return nil
  }

  private func attach(_ scene: SceneLayer) {
    switch scene.content {
    case .camera(let camera): camera.attach(to: self)
    case .image(let image): image.attach(to: self)
    }
  }

  private func detach(_ scene: SceneLayer) {
    switch scene.content {
    case .camera(let camera): camera.detach(from: self)
    case .image(let image): image.detach(from: self)
    }
  }

  private func updateCameraLayerFlag() {
    let attached = videoMixer.hasCameraLayer
    lock.lock()
    cameraLayerAttached = attached
    lock.unlock()
  }

  private func setAudioSourceInternal(_ source: HybridMicrophoneSource?) {
    audioLock.lock()
    let previous = microphone
    microphone = source
    audioLock.unlock()
    if let previous, previous !== source {
      previous.detach(from: self)
    }
    source?.attach(to: self)
  }

  /// Host clock seconds to the session's millisecond clock. The first
  /// frame handed to the video encoder defines t=0; a wall-clock epoch taken at
  /// `startEncoding` would clamp every already-captured frame to 0 and hand the
  /// muxer duplicate timestamps.
  private func ptsMs(for time: CFTimeInterval) -> UInt32 {
    lock.lock()
    if epoch == nil { epoch = time }
    let epoch = self.epoch ?? time
    lock.unlock()
    return HybridMixer.milliseconds(time - epoch)
  }

  /// One ADTS packet from the AAC encoder (microphone queue). While a camera
  /// layer is attached and no video frame has reached the session yet, packets
  /// are kept (not dropped) so audio starts within one packet of the video;
  /// `videoFrameEncoded` flushes them. Packets captured before the epoch (the
  /// first video frame) are dropped. Without a camera layer the first audio
  /// packet defines the epoch.
  private func audioPacketEncoded(_ adts: Data, time: CFTimeInterval) {
    lock.lock()
    if cameraLayerAttached && !videoStarted {
      pendingAudio.append((adts: adts, time: time))
      if pendingAudio.count > HybridMixer.maxPendingAudio { pendingAudio.removeFirst() }
      lock.unlock()
      return
    }
    if epoch == nil { epoch = time }
    pushAudioLocked(adts, time: time)
    lock.unlock()
  }

  /// Caller holds `lock` and `epoch` is set. A packet that ended before the
  /// epoch is dropped; the one spanning it is kept at pts 0, so audio starts
  /// within one packet of the video.
  private func pushAudioLocked(_ adts: Data, time: CFTimeInterval) {
    guard let epoch, let sink else { return }
    let packetDuration = 1024 / max(audioSettings.sampleRate, 8000)
    guard time + packetDuration > epoch else { return }
    encodedAudioFrames += 1
    sink.pushEncodedAudio(adts, ptsMs: HybridMixer.milliseconds(time - epoch))
  }

  private static func milliseconds(_ seconds: CFTimeInterval) -> UInt32 {
    let ms = seconds * 1000
    guard ms.isFinite, ms > 0 else { return 0 }
    return ms >= Double(UInt32.max) ? UInt32.max : UInt32(ms.rounded())
  }

  /// Encoder output thread, after the frame was posted to the session: the
  /// first frame releases the audio packets kept in `audioPacketEncoded`.
  private func videoFrameEncoded(bytes: Int) {
    let now = CACurrentMediaTime()
    lock.lock()
    encodedVideoFrames += 1
    if !videoStarted {
      videoStarted = true
      for packet in pendingAudio { pushAudioLocked(packet.adts, time: packet.time) }
      pendingAudio.removeAll()
    }
    outputWindow.append((time: now, bytes: bytes))
    outputWindow.removeAll { now - $0.time > 1 }
    lock.unlock()
  }

  private func reportEncoderFailure(_ message: String) {
    lock.lock()
    encoderFailures += 1
    let callback = errorCallback
    lock.unlock()
    callback?(CaptureError(code: .encoderfailed, message: message))
  }
}

File: android/src/main/java/com/margelo/nitro/nitrortmp/HybridMixer.kt
package com.margelo.nitro.nitrortmp

import android.util.Log
import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.nitrortmp.capture.MicrophoneSource
import com.margelo.nitro.nitrortmp.encoder.AacEncoder
import com.margelo.nitro.nitrortmp.encoder.H264Encoder
import com.margelo.nitro.nitrortmp.mixer.VideoMixer
import java.util.ArrayDeque
import java.util.concurrent.atomic.AtomicLong
import kotlin.math.roundToInt

/** The `MixerStats` counters, written by the render, encoder and audio threads. */
internal class MixerCounters {
  val capturedFrames = AtomicLong()
  val renderedFrames = AtomicLong()
  val droppedFrames = AtomicLong()
  val encodedVideoFrames = AtomicLong()
  val encodedAudioFrames = AtomicLong()
  val encoderFailures = AtomicLong()
  @Volatile var videoBitrateKbps = 0.0
}

/** What `CameraLayer` and `ImageLayer` implement so the mixer can hook them up. */
internal interface MixerLayer {
  fun attach(mixer: HybridMixer)
  fun detach()
}

/**
 * The `Mixer` HybridObject: the scene (layers, z order,
 * rectangles), the output format, the compositor ([VideoMixer]) and both
 * encoders. Encoding starts when the attached session reaches `publishing`
 * (`startEncoding`) and stops with it; the preview keeps running
 * without a session.
 *
 * Threading: JS calls post to the render thread; encoder output comes
 * on the codec's thread and audio output on the microphone thread, both
 * only post to the session through [EncodedFrameSink]. `stats` and
 * `isEncoding` read atomics.
 */
@DoNotStrip
@Keep
class HybridMixer : HybridMixerSpec() {
  internal val counters = MixerCounters()
  internal val videoMixer = VideoMixer(counters)

  private val lock = Any()
  private var videoSettings = VideoSettings(720.0, 1280.0, 30.0, 2500.0, 2.0)
  private var audioSettings = AudioSettings(48000.0, 1.0, 128.0)
  private val layers = ArrayList<HybridVideoLayerSpec>()
  private var audioSource: HybridMicrophoneSource? = null
  @Volatile private var errorCallback: ((CaptureError) -> Unit)? = null

  // Encoding state: written on the render thread, read on the encoder/audio threads.
  @Volatile private var encoding = false
  @Volatile private var sink: EncodedFrameSink? = null
  @Volatile private var epochNs = 0L
  @Volatile private var firstVideoPushed = false
  /** ADTS packets + capture time (µs) encoded before the first video frame went out; guarded by lock. */
  private val pendingAudio = ArrayDeque<Pair<ByteArray, Long>>()
  private var h264: H264Encoder? = null
  private val audioLock = Any()
  private var aac: AacEncoder? = null
  private val bitrateWindow = ArrayDeque<LongArray>()  // [ms, bytes], last second, guarded by lock

  init {
    videoMixer.setOutput(720, 1280, 30.0)
    videoMixer.onEncoderSurfaceFailed = { message ->
      counters.encoderFailures.incrementAndGet()
      reportError(CaptureError(CaptureErrorCode.ENCODERFAILED, message))
    }
  }

  // --- HybridMixerSpec ---------------------------------------------------------------

  override var video: VideoSettings
    get() = synchronized(lock) { videoSettings }
    set(value) {
      synchronized(lock) { videoSettings = value }
      // Only the next startEncoding picks it up while encoding (spec).
      if (!encoding) videoMixer.setOutput(value.width.toInt(), value.height.toInt(), value.frameRate)
    }

  override var audio: AudioSettings
    get() = synchronized(lock) { audioSettings }
    set(value) = synchronized(lock) { audioSettings = value }

  override val stats: MixerStats
    get() = MixerStats(
      capturedFrames = counters.capturedFrames.get().toDouble(),
      renderedFrames = counters.renderedFrames.get().toDouble(),
      droppedFrames = counters.droppedFrames.get().toDouble(),
      encodedVideoFrames = counters.encodedVideoFrames.get().toDouble(),
      encodedAudioFrames = counters.encodedAudioFrames.get().toDouble(),
      videoBitrateKbps = counters.videoBitrateKbps,
      encoderFailures = counters.encoderFailures.get().toDouble(),
    )

  override val isEncoding: Boolean
    get() = encoding

  override fun addLayer(layer: HybridVideoLayerSpec, frame: LayerFrame?) {
    synchronized(lock) {
      if (layers.any { it === layer }) return
      layers.add(layer)
    }
    // The slot exists on the render thread before the layer may ask for a surface.
    videoMixer.addLayer(layer, frame)
    (layer as? MixerLayer)?.attach(this)
  }

  override fun removeLayer(layer: HybridVideoLayerSpec) {
    synchronized(lock) {
      if (!layers.removeIf { it === layer }) return
    }
    (layer as? MixerLayer)?.detach()
    videoMixer.removeLayer(layer)
  }

  override fun setLayerFrame(layer: HybridVideoLayerSpec, frame: LayerFrame) {
    videoMixer.setLayerFrame(layer, frame)
  }

  override fun setAudioSource(source: HybridMicrophoneSourceSpec?) {
    val next = source as? HybridMicrophoneSource
    val previous = synchronized(lock) {
      val p = audioSource
      audioSource = next
      p
    }
    if (previous === next) return
    previous?.detach()
    next?.attach(this)
  }

  override fun onError(callback: (error: CaptureError) -> Unit) {
    errorCallback = callback
  }

  override fun dispose() {
    stopEncoding()
    val (attached, mic) = synchronized(lock) {
      val copy = ArrayList(layers)
      layers.clear()
      val m = audioSource
      audioSource = null
      copy to m
    }
    for (layer in attached) (layer as? MixerLayer)?.detach()
    mic?.detach()
    videoMixer.release()
    super.dispose()
  }

  // --- session link --------------------------------------------------------------

  /** Called by the session on its queue; idempotent. */
  internal fun startEncoding(sink: EncodedFrameSink) {
    videoMixer.handler.post { startOnRender(sink) }
  }

  /** Called by the session on its queue; idempotent. */
  internal fun stopEncoding() {
    videoMixer.handler.post { stopOnRender() }
  }

  private fun startOnRender(sink: EncodedFrameSink) {
    if (encoding) return
    val v = video
    val a = audio
    videoMixer.setOutput(v.width.toInt(), v.height.toInt(), v.frameRate)
    val encoder = try {
      H264Encoder(
        width = v.width.toInt(),
        height = v.height.toInt(),
        frameRate = v.frameRate.roundToInt().coerceAtLeast(1),
        bitrateKbps = v.bitrateKbps.roundToInt().coerceAtLeast(1),
        keyframeIntervalSeconds = v.keyframeIntervalSeconds ?: 2.0,
        listener = videoListener,
      )
    } catch (e: Exception) {
      counters.encoderFailures.incrementAndGet()
      reportError(CaptureError(CaptureErrorCode.ENCODERFAILED, "H.264 encoder: ${e.message}"))
      return
    }
    try {
      videoMixer.setEncoderSurface(encoder.inputSurface)
    } catch (e: RuntimeException) {
      encoder.release()
      counters.encoderFailures.incrementAndGet()
      reportError(CaptureError(CaptureErrorCode.ENCODERFAILED, "encoder surface: ${e.message}"))
      return
    }
    h264 = encoder
    encoder.requestKeyframe()  // the first frame is a keyframe
    val rate = a.sampleRate.roundToInt()
    if (rate != MicrophoneSource.SAMPLE_RATE) {
      Log.w(TAG, "audio.sampleRate $rate is not supported by the capture; encoding at ${MicrophoneSource.SAMPLE_RATE}")
    }
    synchronized(audioLock) {
      aac = try {
        AacEncoder(
          sampleRate = MicrophoneSource.SAMPLE_RATE,
          channels = a.channels.roundToInt().coerceIn(1, 2),
          bitrateKbps = a.bitrateKbps.roundToInt().coerceAtLeast(8),
          listener = audioListener,
        )
      } catch (e: Exception) {
        counters.encoderFailures.incrementAndGet()
        reportError(CaptureError(CaptureErrorCode.ENCODERFAILED, "AAC encoder: ${e.message}"))
        null
      }
    }
    synchronized(lock) { bitrateWindow.clear() }
    counters.videoBitrateKbps = 0.0
    epochNs = 0L  // established by the first encoded frame
    firstVideoPushed = false
    synchronized(lock) { pendingAudio.clear() }
    this.sink = sink
    encoding = true
  }

  private fun stopOnRender() {
    if (!encoding) return
    encoding = false
    sink = null
    videoMixer.setEncoderSurface(null)
    h264?.release()
    h264 = null
    synchronized(audioLock) {
      aac?.release()
      aac = null
    }
    counters.videoBitrateKbps = 0.0
  }

  // --- media input ---------------------------------------------------------------------------

  /** PCM from the attached microphone, on its thread; encoded synchronously. */
  internal fun audioArrived(pcm: ShortArray, frames: Int, sampleRate: Int, channels: Int, timestampNs: Long) {
    if (!encoding) return
    synchronized(audioLock) {
      aac?.encode(pcm, frames, channels, timestampNs)
    }
  }

  internal fun reportError(error: CaptureError) {
    errorCallback?.invoke(error)
  }

  private fun hasCameraLayer(): Boolean = synchronized(lock) { layers.any { it is HybridCameraLayer } }

  private val videoListener = object : H264Encoder.Listener {
    override fun onEncodedVideo(annexb: ByteArray, presentationTimeUs: Long, keyframe: Boolean) {
      if (!encoding) return
      val s = sink ?: return
      // Under the lock so no audio packet slips into the session before the first video frame.
      synchronized(lock) {
        val epochUs = establishEpochUs(presentationTimeUs)
        val pts = ((presentationTimeUs - epochUs) / 1000).coerceAtLeast(0)
        s.pushEncodedVideo(annexb, pts.toInt(), pts.toInt())
        if (!firstVideoPushed) {
          firstVideoPushed = true
          for ((adts, timeUs) in pendingAudio) pushAudioLocked(s, adts, timeUs, epochUs)
          pendingAudio.clear()
        }
      }
      counters.encodedVideoFrames.incrementAndGet()
      updateBitrate(annexb.size.toLong())
    }

    override fun onEncoderError(message: String) {
      counters.encoderFailures.incrementAndGet()
      reportError(CaptureError(CaptureErrorCode.ENCODERFAILED, message))
    }
  }

  private val audioListener = object : AacEncoder.Listener {
    override fun onEncodedAudio(adts: ByteArray, presentationTimeUs: Long) {
      if (!encoding) return
      val s = sink ?: return
      synchronized(lock) {
        // Audio waits for the first video frame so the stream starts on the
        // keyframe and both clocks share the epoch; the packets are kept,
        // not dropped, so audio starts within one packet of the video.
        if (!firstVideoPushed && hasCameraLayer()) {
          pendingAudio.addLast(adts to presentationTimeUs)
          while (pendingAudio.size > MAX_PENDING_AUDIO) pendingAudio.removeFirst()
          return
        }
        pushAudioLocked(s, adts, presentationTimeUs, establishEpochUs(presentationTimeUs))
      }
    }

    override fun onEncoderError(message: String) {
      counters.encoderFailures.incrementAndGet()
      reportError(CaptureError(CaptureErrorCode.ENCODERFAILED, message))
    }
  }

  /**
   * Pushes a packet unless it ended before the epoch (the first video frame).
   * The packet that spans the epoch is kept at pts 0, so audio starts within
   * one packet of the video. Caller holds lock.
   */
  private fun pushAudioLocked(s: EncodedFrameSink, adts: ByteArray, timeUs: Long, epochUs: Long) {
    if (timeUs + AAC_PACKET_US <= epochUs) return
    val pts = ((timeUs - epochUs) / 1000).coerceAtLeast(0)
    s.pushEncodedAudio(adts, pts.toInt())
    counters.encodedAudioFrames.incrementAndGet()
  }

  /**
   * The session clock's origin: the capture time of the first frame that
   * reaches the sink, not the wall clock at `startEncoding`. Camera frames
   * captured before `startEncoding` would otherwise all clamp to 0 and give
   * the muxer duplicate timestamps.
   */
  private fun establishEpochUs(presentationTimeUs: Long): Long = synchronized(lock) {
    if (epochNs == 0L) epochNs = presentationTimeUs * 1000
    epochNs / 1000
  }

  /** Bytes over the last second, from the encoder thread. */
  private fun updateBitrate(bytes: Long) {
    val now = System.nanoTime() / 1_000_000
    val kbps: Double
    synchronized(lock) {
      bitrateWindow.addLast(longArrayOf(now, bytes))
      while (bitrateWindow.isNotEmpty() && now - bitrateWindow.peekFirst()!![0] > 1000) bitrateWindow.pollFirst()
      var sum = 0L
      for (entry in bitrateWindow) sum += entry[1]
      kbps = sum * 8.0 / 1000.0
    }
    counters.videoBitrateKbps = kbps
  }

  companion object {
    private const val TAG = "nitrortmp.mixer"
    /** About 2 s of AAC packets kept while waiting for the first video frame. */
    private const val MAX_PENDING_AUDIO = 100
    /** One AAC packet: 1024 samples at the capture rate. */
    private const val AAC_PACKET_US = 1024L * 1_000_000L / MicrophoneSource.SAMPLE_RATE
  }
}
